haystack/nodes/prompt/prompt_node.py

```diff
@@ -42,60 +52,203 @@ def run_batch(
         raise NotImplementedError("This method should never be implemented in the derived class")
 
 
+PROMPT_TEMPLATE_ALLOWED_FUNCTIONS = ast.literal_eval(
+    os.environ.get(HAYSTACK_PROMPT_TEMPLATE_ALLOWED_FUNCTIONS, '["join", "to_strings", "replace", "enumerate", "str"]')
```

Comment:
@tholor as you initially came up with the idea to support f-string like syntax, please have a look at the new syntax in general and especially on those default allowed functions. Do you think we're missing something essential? 
Note that this list can be changed in any case via env variable. Also note that we can only filter by function/method name here, but not on the instance type. So `join` here enables two-fold: 
- our own `join` shaping function we implement in shaper.py as `join_documents` and publish it to the globals as `join` via `import ... as join` here.
- string's `join` method such as `'delim'.join([d.content for d in documents])`

Reply:
I got that we use per default `join_documents` under the hood if a user puts `join()` into the template. 
What I didn't get: 
- What do you mean in the above comment with "string's join method such as 'delim'.join([d.content for d in documents])"?
- Do you suggest that the user can register a different "join" function by importing it? How would that look like in an example?

Reply:
I really like the flexibility of using the regular string's `join` and list comprehensions within a prompt template.
From the top of my head, I am not missing any particular allowed function here and it should be relatively easy to extend them later on. 
In any case, I'd like to see **excellent** documentation on those allowed functions and examples on what each function actually does. An example where we join documents via a regular python list comprehension and string's join method + then the shorter alternative via our "internal" join_documents function could help a lot to understand this. 

Despite the syntax: I see one potential security concern with the current implementation, that I would like to double-check with you: 
- We basically allow calling the `join()` method of any arbitrary python object
- So if someone passes a custom object to the template with a `join()` method they can basically run any code that is defined in this method (see code example below)
- Good news: Users still won't be able inject any code via a custom prompt template supplied to an externally running pipeline (e.g. dC), but **only** if they control the input object
- If users can only pass custom objects to a promptnode in situations where they could run arbitrary python functions anyway already on that machine, this won't make any difference and we are good
- Do you see any cases where an attacker could pass a custom object to a pipeline running in a different python interpreter? 

```python
import os
from haystack.nodes.prompt import PromptTemplate

os.environ["TEST_SECRET"] = "SOME_SECRECT_VALUE"

class Hijack():

  def join(self):
    # run any python code
    with open("some.txt", 'w') as f:
      f.write("some value")
      
    # return any env variable
    return os.getenv("TEST_SECRET")


pt = PromptTemplate(name="give-a-title", prompt_text="Some prompt text {documents.join()}")
hijack = Hijack()

print(list(pt.fill(documents=hijack)))

# returns: 
# ['Some prompt text SOME_SECRECT_VALUE']

```

Reply:
> Do you see any cases where an attacker could pass a custom object to a pipeline running in a different python interpreter?
I don't think it's possible without importing additional code and then the malicious code could already run at import time.